### ISSUE TITLE
Failed to reproduce the intro.ipynb

### ISSUE THREAD

**Issue by ChrisW-priv** [stanfordnlp/dspy#1775]
Hi! 
I will start by saying that I love the idea behind DSPy. Really interesting project and nice idea in general. 

However, when I was trying to learn the framework I started copy pasting the code of the intro.ipynb and created attached file. 

As you will see when you scroll down the result of baleen compilation (or even the uncompiled version) are complete oposite to the ones in the intro. 

Results I got are:

## Retrieval Score for compiled Baleen: 2.0

where as the one in the intro was 

## Retrieval Score for compiled Baleen: 60.0

would you mind taking a look and telling me if this is my error or "it just happens sometimes"? I really thought the results would be the same (using the same model + temp=0), code itself was just copy-pasted (without the md parts) 

Thanks in advance!

Below is the link to the converted .ipynb file to .md cuz I could not upload any other.

[Dspy tutorial.md](https://github.com/user-attachments/files/17676919/Dspy.tutorial.md)

**Comment by okhat** [comment]
Hey @ChrisW-priv ! Thanks for the detailed issue.

First, this entirely on us for pinning this notebook to DSPy 2.4 (the top of the file pins `!pip install dspy-ai==2.4.17`) and not updating it since the original `gpt-3.5-turbo` (Jun 2023) was discontinued and switched to a different (Jan 2024) version, while we migrate the documentation to 2.5.

Based on your file, you unpinned and ran this on 2.5, which behaves very differently due to the introduction of [Adapters](https://github.com/stanfordnlp/dspy/blob/main/examples/migration.ipynb). FWIW, the [two intro notebooks for 2.5 are here instead](https://dspy-docs.vercel.app/quick-start/getting-started-01/).

That said, let's work through this to understand what you observed in 2.5 land. In your logs, you can see that the teleprompter call outputs:

> Bootstrapped 0 full traces after 20 examples in round 0.

This means that none of the examples passed the filter, so it wasn't able to build examples! Why might that happen? Looking at inspect history, you can see that the model produces SQL queries, not search queries. If you use a powerful optimizer in DSPy, like MIPROv2 I'd wager it'd be able with enough training data to move even a small/stubborn model back into generating search queries.

That said, since the issue is qualitative, we can just adjust the field and call it `search_query`. And edit the `.query` to `.search_query` in the `forward` method of `SimplifiedBaleen` as well.

```python
class GenerateSearchQuery(dspy.Signature):
    """Write a simple search query that will help answer a complex question."""

    context = dspy.InputField(desc="may contain relevant facts")
    question = dspy.InputField()
    search_query = dspy.OutputField()
```


Running with this, I get 46% and 52% on `gold_passages_retrieved` with `SimplifiedBaleen` before and after `BootstrapFewShot`, respectively. Further changing the teleprompter call to simplify the metric as follows `dspy.BootstrapFewShot(metric=exact_match, max_labeled_demos=4)` gives us 64% after compiling.

This is all on 50 dev examples, though, so you shouldn't read too much into the exact numbers. This notebook, originating in January 2023 and updated in Aug 2023, just illustrates the very basic workflow in a very cheap setting that ensures no users overspend back when models were more expensive, rather than serving as a complete application. We've since moved to illustrating more complete workflows, using actual DSPy optimizers like BootstrapFewShotWithRandomSearch or MIPROv2 on more substantial amounts of data in the 2.5 documentation: https://dspy-docs.vercel.app/quick-start/getting-started-01/

In any case, I'll follow up beneath this with a complete code snippet upgraded to `gpt-4o-mini` for your reference in DSPy 2.5.

**Comment by okhat** [comment]
Here's a snippet in DSPy 2.5 with GPT-4o-mini. With all the usual disclaimers about the tiny data size in this illustrative notebook, the snippet below uses MIPROv2 in `light` mode. Before optimization, it scores 24% and 58% on evaluating the answer and retrieval, respectively. After optimization, these become 52% and 64%.

```python
import dspy
from dspy.datasets import HotPotQA

turbo = dspy.LM('openai/gpt-4o-mini')
colbert = dspy.ColBERTv2(url='http://20.102.90.50:2017/wiki17_abstracts')
dspy.configure(lm=turbo, rm=colbert)

dataset = HotPotQA(train_seed=1, train_size=20, eval_seed=2023, dev_size=50, test_size=0)
trainset = [x.with_inputs('question') for x in dataset.train]
devset = [x.with_inputs('question') for x in dataset.dev]

evaluate = dspy.Evaluate(devset=devset, num_threads=24, display_progress=True, display_table=5)
exact_match = dspy.evaluate.answer_exact_match

def gold_passages_retrieved(example, pred, trace=None):
    gold_titles = set(map(dspy.evaluate.normalize_text, example['gold_titles']))
    found_titles = set(map(dspy.evaluate.normalize_text, [c.split(' | ')[0] for c in pred.context]))
    return gold_titles.issubset(found_titles)

from dsp.utils import deduplicate

class SimplifiedBaleen(dspy.Module):
    def __init__(self, passages_per_hop=3, max_hops=2):
        self.generate_query = dspy.ChainOfThought("context, question -> search_query")
        self.retrieve = dspy.Retrieve(k=passages_per_hop)
        self.generate_answer = dspy.ChainOfThought("context, question -> answer")
        self.max_hops = max_hops
    
    def forward(self, question):
        context = []
        
        for hop in range(self.max_hops):
            query = self.generate_query(context=context, question=question).search_query
            context = deduplicate(context + self.retrieve(query).passages)

        pred = self.generate_answer(context=context, question=question)
        return dspy.Prediction(context=context, answer=pred.answer)

tp = dspy.MIPROv2(metric=exact_match, auto="light", num_threads=8)
optimized_react = tp.compile(SimplifiedBaleen(), trainset=trainset, requires_permission_to_run=False)

for metric in [exact_match, gold_passages_retrieved]:
    evaluate(SimplifiedBaleen(), metric=metric)
    evaluate(optimized_react, metric=metric)
```

**Comment by ChrisW-priv** [comment]
thanks, for the quick and verbose reply! will look into all that you said and will post results when I am done :-)

**Comment by ChrisW-priv** [comment]
@okhat  I would say that from perspective of the user, the second error is quite weird as the assert refers to internal params of the algorithm (which for un-accustomed users are just not known).

That being said, I am now reading more on the algorithm and stuff so this is a skill issue on my part mostly 😄

**Comment by okhat** [comment]
Hey @ChrisW-priv , are you on an old version by any chance?

These errors shouldn't happen if you upgrade: `pip install -U dspy`

**Comment by ChrisW-priv** [comment]
For a second I could not undestand what you meant as I posted you the versions... or at least I though I did untill I saw what I attached... 😅

My dspy versions I used during the coding:

```
dspy==2.5.22
dspy-ai==2.5.4
```

I just run the upgrade with `uv` I got the 

```
dspy==2.5.29
dspy-ai==2.5.29
```

I tried running the code and now it works.

Sorry for the trouble etc. Now, that being said, the versioning is now the best... change from `2.5.22` to `2.5.29` indicates minor patch. 

But I am not here to complain haha, glad it works and thanks for the tips so far. I will be sure to reach out if I have some more questions. Have a great weekend :-)
